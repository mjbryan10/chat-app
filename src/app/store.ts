import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import loginReducer from '../features/login/loginSlice';
import usersReducer from '../features/users/usersSlice';
import themeReducer from '../features/theme/themeSlice';
import conversationReducer from '../features/conversation/conversationSlice';
import messageReducer from '../features/message/messageSlice';
import navigationReducer from '../features/navigation/navigationSlice'

export const store = configureStore({
  reducer: {
    login: loginReducer,
    users: usersReducer,
    theme: themeReducer,
    conversation: conversationReducer,
    message: messageReducer,
    navigation: navigationReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

/**
 * String values available to the loading status.
 */
export type LoadingStatus = 'rejected' | 'pending' | 'fulfilled' | 'idle';
