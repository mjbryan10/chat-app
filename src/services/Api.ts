const Api = {
  baseUrl: 'http://assignment.bunq.com'
}

export default Api;