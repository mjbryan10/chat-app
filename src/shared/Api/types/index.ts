export interface User {
   id: number;
   name: string;
}

export interface LoginResponse extends User {
   isAuthenticated: boolean;
}

export interface ConversationDetails {
   id: number | null;
   conversationId: number | null;
   is_owner: number;
   userid: number;
   status: number | null;
   lastseen: string | null;
   name: string | null;
   type: number | null;
}

export interface ConversationUser {
   id: number;
   conversationId: number;
   is_owner: number;
   userid: number;
   status: number | null;
   lastseen: string | null;
}

export interface Conversation {
   conversation: ConversationDetails;
   users: ConversationUser[];
}

export interface Message {
   id: number;
   senderId: number;
   message: string;
   timestamp: string;
   conversationid: number;
   status: number | null;
}

export type NewMessagesResponse = Message[] | { message: string };

export interface ApiPostResponse {
   id: number;
}

export const TIME_FORMAT = 'YYYY-MM-DD hh:mm:ss'