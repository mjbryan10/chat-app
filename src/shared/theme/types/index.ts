/**
 * Class for making an array of colors
 * 
 * 
 * Technique to avoid hard typing names into array:
 * @see https://stackoverflow.com/a/59806829/12873927
 */
class ColorSpectrumList {
   /**
    * 
    */
   constructor(
      readonly blue?: string,
      readonly red?: string,
      readonly green?: string,
      readonly purple?: string,
      readonly orange?: string,
      readonly pink?: string,
      readonly indigo?: string,
      readonly teal?: string,
      readonly yellow?: string,
      readonly cyan?: string,
   ){}
}
export type SpectrumColor = keyof ColorSpectrumList;
/**
 * An array of color names for use.
 */
export const colorSpectrumArray = Object.keys(new ColorSpectrumList()) as SpectrumColor[]; 
/**
 * Color Spectrum
 * 
 * Implements class and makes fields no longer optional.
 */
export interface ColorSpectrum extends ColorSpectrumList {
   blue: string;
   indigo: string;
   purple: string;
   pink: string;
   red: string;
   orange: string;
   yellow: string;
   green: string;
   teal: string;
   cyan: string;
   /**
    * Color for the owner of message/conversation
    */
   owner: string;
}
/**
 * Colors found on the theme
 */
export interface ThemeColors extends ColorSpectrum {
   primary: string;
   secondary: string;
   success: string;
   warning: string;
   danger: string;
   info: string;
   dark: string;
   light: string;
   text: string;
   white: string;
   gray: string;
   grayDark: string;
}
/**
 * Possible colors available for use in theme
 */
export type ThemeColor = keyof ThemeColors;
/**
 * Background styles found on the theme
 */
export interface ThemeBackgrounds {
   page: string;
   message: string;
   buttonPrimary: string;
   inputField: string;
   conversation: string;
   conversationSelected: string;
}

/**
 * Export of Theme to 'styled-components'
 */
declare module 'styled-components' {
   /**
    * Theme interface
    */
   export interface Theme {
      color: ThemeColors;
      background: ThemeBackgrounds;
      border: string;
   }
   /**
    * Possible colors available for use in theme
    */
   export type ThemeColor = keyof ThemeColors;
}
/**
 * Theme types
 */
export type ThemeName = 'dark' | 'light';

/**
 * Checks if given value is a valid theme name.
 *
 * Returns `true` if given value is a valid theme name.
 *
 * Returns `false` if given value is not a valid theme name.
 * @param value Any given value to check if matches theme names
 */
export function isThemeName(value: any): value is ThemeName {
   if (value === 'light' || value === 'dark') {
      return true;
   }
   return false;
}
