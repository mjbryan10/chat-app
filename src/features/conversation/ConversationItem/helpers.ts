import { Conversation } from "shared/Api/@types";
import UserApi from "shared/Api/UserApi";

/**
 * Returns the conversation name if it is recorded.
 * 
 * If the name is not recorded, searches the coversation for another user,
 * and returns that users name if that is available.
 * 
 * If unable to find users name from API returns 'Anonymous'
 * 
 * Default returns 'Unknown conversation name' if all else fails.
 * @param conversation The conversation the name relates to
 * @param currentUserId The current user requesting the conversation name
 */
export const computeConversationName = async (
  conversation: Conversation,
  currentUserId: number | null
): Promise<string> => {
  const { name } = conversation.conversation;

  if (name) return name;

  const friendUser = conversation.users.find((user) => user.userid !== currentUserId);

  if (friendUser) {
    try {
      return await new UserApi().fetchUserNameById(friendUser.userid)
    } catch (error) {
      return 'Anonymous'
    } //TODO: Test if this works!
    
    //  const response = await new UserApi().fetchUserNameById(friendUser.userid)
     

    //  return response.data ? response.data.name : 'Anonymous';
  }
  return 'Unknown conversation name';
};
/**
 * Compares a conversation to the one in the Redux store and returns `true` if this conversation
 * is the selected on in the redux store.
 * 
 * Returns `false` if otherwise.
 * @param conversation the conversation to compare to selected conversation
 * @param currentConversationId The ID of the current conversation selected in Redux store
 */
export const isSelectedConversation = (
  conversation: Conversation,
  currentConversationId: number | null
): boolean => {
  if (!currentConversationId) return false;
  if (conversation.conversation.conversationId === currentConversationId) return true;
  return false;
};