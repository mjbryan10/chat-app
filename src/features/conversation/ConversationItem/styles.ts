import styled, { ThemeProps, Theme } from 'styled-components';
import { SpectrumColor } from 'theme/@types';

interface ButtonProps extends ThemeProps<Theme> {
  selected: boolean;
//   color: SpectrumColor;
}

export const ItemButton = styled.button<ButtonProps>`
   &,
   &:active {
      background: ${({ theme: { background }, selected }) =>
         selected ? background.conversationSelected : background.conversation};
   }
`;
   /* border: ${({ color, theme }) => theme.color[color]} 1px solid */
