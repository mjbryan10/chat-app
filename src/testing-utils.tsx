import React from 'react';
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { dark } from './shared/theme';

const AllTheProviders = ({ children }: any) => {
  return (
    <ThemeProvider theme={dark}>
        {children}
    </ThemeProvider>
  )
  
}

const customRender = (ui?: any, options?: any) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }