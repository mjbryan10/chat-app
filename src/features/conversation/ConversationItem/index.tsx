import React, { SyntheticEvent, FC, useMemo, useEffect, useState } from 'react';
import * as S from './styles';
import AvatarWithLabel from 'components/AvatarWithLabel';
import { Conversation } from 'shared/Api/types';
import { useSelector, useDispatch } from 'react-redux';
import { selectLogin } from 'features/login/loginSlice';
import {
   selectCurrentConversationId,
   setCurrentConversation,
} from '../conversationSlice';
import { isSelectedConversation, computeConversationName } from './helpers';
import { SpectrumColor } from 'shared/theme/types';

interface Props {
   conversation: Conversation;
   color: SpectrumColor;
}
/**
 * A React functional component for rendering a conversation option.
 * 
 * When the component is clicked by the user it will update the current conversation in
 * the Redux store.
 * @param conversation The conversation to which this card item relates to.
 */
const ConversationItem: FC<Props> = ({ conversation, color }) => {
   const { id: currentUserId } = useSelector(selectLogin);
   const currentConversationId = useSelector(selectCurrentConversationId);
   const dispatch = useDispatch();
   const { id } = conversation.conversation;

   /**
    * Calculated boolean value which equates to true should this conversation
    * match the currentConversationId
    */
   const selected = useMemo(
      () => isSelectedConversation(conversation, currentConversationId),
      [conversation, currentConversationId]
   );
   /**
    * Local state for holding conversationName
    */
   const [conversationName, setConversationName] = useState('Loading..');
   /**
    * Handles the computation and updating of the conversation name.
    * 
    * Will re-render and re-compute conversation name should the conversation or
    * currentUserId props change.
    */
   useEffect(() => {
      computeConversationName(conversation, currentUserId).then((name) => {
         setConversationName(name);
      });
   }, [conversation, currentUserId]);

   /**
    * On button click will dispatch and event to update the current conversation
    * in Redux store to this conversation
    *
    * @param event HTMLButton Click Event
    */
   const onClick = (event: SyntheticEvent<HTMLButtonElement>) => {
      event.preventDefault();
      dispatch(setCurrentConversation(id));
   };
   return (
      <S.ItemButton
         selected={selected}
         onClick={onClick}
         data-testid="conversation-item-button"
         color={color}
      >
         <AvatarWithLabel title={conversationName} key={id} color={color}  />
      </S.ItemButton>
   );
};

export default ConversationItem;
