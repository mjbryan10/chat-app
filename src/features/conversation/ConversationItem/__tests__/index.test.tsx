import React from 'react';
import { render, cleanup, fireEvent } from 'testing-utils';
import ConversationItem from '..';
import { Conversation } from 'shared/Api/types';
import axios from 'axios';
import { Provider } from 'react-redux';
import { store } from 'app/store';
import { act } from 'react-dom/test-utils';

describe('AvatarWithLabelAndDelete component', () => {
   const privateConversation: Conversation = {
      conversation: {
         id: 4010,
         conversationId: 4010,
         is_owner: 0,
         userid: 1,
         status: 1,
         lastseen: null,
         name: null,
         type: 1,
      },
      users: [
         {
            id: 10257,
            conversationId: 4010,
            is_owner: 0,
            userid: 1,
            status: null,
            lastseen: null,
         },
         {
            id: 10258,
            conversationId: 4010,
            is_owner: 0,
            userid: 2,
            status: null,
            lastseen: null,
         },
      ],
   };
   const groupConversation: Conversation = {
      conversation: {
         id: 4,
         conversationId: 4,
         is_owner: 0,
         userid: 1,
         status: 1,
         lastseen: '2019-11-05 10:10:50',
         name: 'Group Chaaat',
         type: 2,
      },
      users: [
         {
            id: 31,
            conversationId: 4,
            is_owner: 0,
            userid: 1,
            status: null,
            lastseen: '2019-11-05 10:10:50',
         },
         {
            id: 32,
            conversationId: 4,
            is_owner: 0,
            userid: 2,
            status: null,
            lastseen: '2019-01-25 11:49:01',
         },
         {
            id: 33,
            conversationId: 4,
            is_owner: 0,
            userid: 3,
            status: null,
            lastseen: null,
         },
         {
            id: 34,
            conversationId: 4,
            is_owner: 0,
            userid: 4,
            status: null,
            lastseen: null,
         },
      ],
   };
   beforeEach(() => {
      cleanup();
      (axios.get as jest.Mock).mockClear();
   });
   // test('should fire clickHandler', () => {
   //    const clickHandler = jest.fn();
   //    const utils = render(
   //       <ConversationItem conversation={privateConversation} title="test title" handleClick={clickHandler} />
   //    );
   //    const button = utils.getByTestId('conversation-item-button');

   //    fireEvent.click(button);

   //    expect(clickHandler).toHaveBeenCalledTimes(1);
   // });
   // test('should render title text as avatar label', () => {
   //    const clickHandler = jest.fn();
   //    const testTitle = 'test title';
   //    const utils = render(
   //       <ConversationItem id={1} title={testTitle} handleClick={clickHandler} />
   //    );

   //    const innerContainer = utils.getByTestId('conversation-item-button');

   //    expect(innerContainer).toHaveTextContent(testTitle);
   // });

   // test('should fetch user name for title if no conversation name', async () => {
   //    (axios.get as jest.Mock).mockImplementationOnce(() =>
   //       Promise.resolve({
   //          id: '1',
   //          name: 'Wessel',
   //       })
   //    );
   //    const utils = render(
   //       <Provider store={store}>
   //          <ConversationItem conversation={privateConversation} color="blue" />
   //       </Provider>
   //    );
   //    const avatar = utils.getByTestId('avatar-inner-container');
   //    const label = utils.getByTestId('avatar-with-label-span');

   //    act(() => {
   //       setTimeout(() => {
   //       }, 1000);
   //    });
   //    expect(avatar).toHaveTextContent('W');
   //    expect(label).toHaveTextContent('Wessel');
   //    expect(axios.get).toBeCalledTimes(1);
   // });
});
