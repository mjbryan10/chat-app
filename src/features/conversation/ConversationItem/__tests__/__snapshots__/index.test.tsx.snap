// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversationItem component should render correctly 1`] = `
<DocumentFragment>
  <button
    class="sc-AxjAm kpMVKU"
    color="red"
    data-testid="conversation-item-button"
  >
    <div
      class="sc-AxirZ knpAQZ"
      color="red"
    >
      <div
        class="sc-AxiKw gVga-dC"
        color="red"
        size="50"
      >
        <span
          class="sc-AxhCb jPjWba"
          data-testid="avatar-inner-container"
          size="50"
        >
          L
        </span>
      </div>
      <h3
        data-testid="avatar-with-label-span"
      >
        Loading..
      </h3>
    </div>
  </button>
</DocumentFragment>
`;
